### ISSUE TITLE
python grpcio 1.9.1: `can't start new thread`

### ISSUE THREAD

**Issue by krelst** [grpc/grpc#14523]
We are using several google-cloud python clients which have grpcio as sub dependency. A few days ago we upgrade the grpcio package from `1.8.4` to `1.9.1` and started seeing stale processes with the 
following error. 

```
Traceback (most recent call last):
  File "/code/pipelines/base_classes/pubsub_pipeline.py", line 133, in process_single_message
    sub.transformer.transform(message_item, transform_result)
  File "/code/transformations/integrity_actions/demand/GroupedIntegritiesTransform.py", line 117, in transform
    aggregate_rate_id_counts(destination_id, extract_date, timestamp=timestamp, persist=True)
  File "/code/demand/rate_id_counts.py", line 174, in aggregate_rate_id_counts
    force_most_recent=force_most_recent_mappings)
  File "/code/demand/demand_mappings.py", line 36, in get_mappings_from_bt
    bt_rows = bt_client.read_rows(DEMAND_MAPPINGS_TABLE, from_key, to_key, filter=CellsColumnLimitFilter(num_cells=1))
  File "/usr/local/lib/python2.7/dist-packages/tenacity/__init__.py", line 214, in wrapped_f
    return self.call(f, *args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/tenacity/__init__.py", line 295, in call
    start_time=start_time)
  File "/usr/local/lib/python2.7/dist-packages/tenacity/__init__.py", line 257, in iter
    trial_time_taken)
  File "/code/utils/bigtableutils/BigtableManager.py", line 24, in reinitialise_bigtable_client
    __bigtable_writer_singletons[bigtable_config["cluster_id"]].create_client()
  File "/code/utils/bigtableutils/BigtableManager.py", line 40, in create_client
    admin=False
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/bigtable/client.py", line 228, in __init__
    self._data_stub = _make_data_stub(self)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/bigtable/client.py", line 98, in _make_data_stub
    extra_options=_GRPC_MAX_LENGTH_OPTIONS)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/_helpers.py", line 590, in make_secure_stub
    extra_options=extra_options)
  File "/usr/local/lib/python2.7/dist-packages/google/cloud/_helpers.py", line 560, in make_secure_channel
    options=options)
  File "/usr/local/lib/python2.7/dist-packages/google/auth/transport/grpc.py", line 135, in secure_authorized_channel
    return grpc.secure_channel(target, composite_credentials, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/grpc/__init__.py", line 1584, in secure_channel
    credentials._credentials)
  File "/usr/local/lib/python2.7/dist-packages/grpc/_channel.py", line 909, in __init__
    _subscribe(self._connectivity_state, lambda *args: None, None)
  File "/usr/local/lib/python2.7/dist-packages/grpc/_channel.py", line 862, in _subscribe
    polling_thread.start()
  File "/usr/lib/python2.7/threading.py", line 736, in start
    _start_new_thread(self.__bootstrap, ())
error: can't start new thread
```

Our setup is the following: we have some pipelines running that are pulling from gcloud pub/sub, processing them and then writing them to gcloud bigtable. These pipelines are deployed as pods on GKE. OS of the Docker image is debian stretch (9.1). Some relevant installed packages:

```
gapic-google-cloud-logging-v2==0.91.3
gapic-google-cloud-pubsub-v1==0.15.4
google-api-python-client==1.6.5
google-auth-httplib2==0.0.3
google-auth==1.4.1
google-cloud-bigquery==0.26
google-cloud-bigtable==0.26
google-cloud-core==0.26.0
google-cloud-logging==1.2.0
google-cloud-pubsub==0.27.0
google-cloud-storage==1.3.0
google-gax==0.15.16
google-resumable-media==0.3.1
googleapis-common-protos[grpc]==1.5.3
grpc-google-iam-v1==0.11.4
grpcio==1.9.1
```

It looks like these errors only occur for long running processes (several hours). I can't yet determine whether it happens in one specific part of the code or at several different places.

**Comment by srini100** [comment]
@krelst is gRPC version the only change? Was it working fine with 1.8.4?

**Comment by krelst** [comment]
It is a curious case. The upgrade of the grpc library took place 16/2, but the first time we saw the error was only 24/2. From then on, multiple different pipelines of ours, which all have grpc/google cloud python clients in common, started failing with the same error. Together with the grpcio upgrade, these other libraries got an upgrade:
```
google-api-python-client==1.6.4 -> 1.6.5
google-auth==1.3.0 -> 1.4.1
```
However it still looks like it is an issue in grpc as we see the following error:
```
"Exception thread.error: error("can't start new thread",) in 'grpc._cython.cygrpc._get_metadata' ignored
```
Of course it could be that some other library is spawning way too much threads, which results in grpc not being able to spawn a new one. But I honestly can't think of any other library that could cause this. We have downgraded grpcio back to 1.8.4, and from then on we haven't see the issue. Will keep you posted if it re-appears with 1.8.4.

**Comment by srini100** [comment]
This issue could be related to [this](https://github.com/grpc/grpc/issues/14578) where large number of timer threads are getting created in C core which doesn't look normal.

**Comment by speedplane** [comment]
Looking through my logs, I see it popped up around February 13, 2018, then went away, and now I am seeing it again. My callstack looks identical to the one posted above. I've now added instrumentation to my code to determine if there are memory leaks or if `threading.active_threads` is growing out of control. 

It seems that with every upgrade to grpc there are major reliability issues. It would be nice if they could take a hard look at their QA process.

**Comment by krelst** [comment]
10 days after downgrading from 1.9.1 to 1.8.4 and haven't seen any issues, so we can be quite certain that the upgrade was the cause of the issue.

**Comment by mehrdada** [comment]
This is just a shot in the dark, but can you please try passing `GRPC_POLL_STRATEGY=poll` environment variable to the current version of gRPC (v1.10.0) and compare results? One thing that changed since 1.8 and 1.9 is the default polling strategy on gRPC Python Linux binary wheels (used to be `poll`, now it is `epoll1`) which improved things on a bunch of dimensions, but that might be the cause of the regressions you're facing?

**Comment by srini100** [comment]
@speedplane, have you determined if this is due to memory leak or too many threads?

**Comment by srini100** [comment]
@krelst @speedplane Any update on this?

**Comment by speedplane** [comment]
@srini100 It's not a memory leak or too many threads. I've instrumented my code to spit out both the amount of free memory and the total number of threads, here are some sample logs:

- `can't start new thread (37.66% mem, 118 threads)`
- `can't start new thread (38.35% mem, 122 threads)`
- `can't start new thread (43.32% mem, 107 threads)`
- `can't start new thread (43.16% mem, 130 threads)`

**Comment by srini100** [comment]
@speedplane, I am assuming you have already checked the thread count limit of a process on your system and that it is much higher than 130. Normally, on a stock 64-bit Linux it is several thousand, although it varies a lot based on a number of settings so it is good to confirm you are not hitting the limit.

Have  you tried GRPC_POLL_STRATEGY=poll suggestion above?

The memory usage seems high. If you have long lived streams you may be hitting [this](https://github.com/grpc/grpc/issues/12383) memory leak bug. I am not sure if max number of threads can change dynamically based on amt of memory consumed by the process. This bug was fixed in 1.10.1 and 1.11.

**Comment by srini100** [comment]
Closing this issue as I have not found any other reason to get this error other than hitting the system limit. The number of threads seem to exceed twice the number of cores. Track the progress of this issue https://github.com/grpc/grpc/issues/14578. I have bumped its priority to get some attention.
